membership-attribute-service/app/controllers/AccountController.scala

```diff
@@ -363,10 +386,11 @@ class AccountController(commonActions: CommonActions, override val controllerCom
   @Deprecated def paperUpdateDirectDebit = updateDirectDebit[SubscriptionPlan.PaperPlan](None)
   def updateDirectDebitOnSpecificSub(subscriptionName: String) = updateDirectDebit[SubscriptionPlan.AnyPlan](Some(memsub.Subscription.Name(subscriptionName)))
 
-
   @Deprecated def membershipDetails = paymentDetails[SubscriptionPlan.PaidMember, SubscriptionPlan.FreeMember]
   @Deprecated def monthlyContributionDetails = paymentDetails[SubscriptionPlan.Contributor, Nothing]
   @Deprecated def digitalPackDetails = paymentDetails[SubscriptionPlan.Digipack, Nothing]
   @Deprecated def paperDetails = paymentDetails[SubscriptionPlan.PaperPlan, Nothing]
+  def allPaymentDetails(productType: Option[String]) = anyPaymentDetails(productType.fold[OptionalSubscriptionsFilter](NoFilter)(FilterByProductType))
```

Comment:
minor comment:
slightly confusing in this fold on first sight is that NoFilter is a value (an object) but FilterByProductType is actually a function.  There's also sometimes an ambiguity caused where the type can't be inferred by the compiler, although clearly not in this case.  I used to strongly oppose fold on an option in favour of pattern match (for readability reasons), but now I'm just prefer to avoid.
```suggestion
  def allPaymentDetails(productType: Option[String]): OptionalSubscriptionsFilter = anyPaymentDetails(productType.fold[OptionalSubscriptionsFilter](NoFilter)(FilterByProductType.apply))
```

Reply:
Conceptually, both `NoFilter` and `FilterByProductType` are _constructors_ of `OptionalSubscriptionsFilter` algebraic data type, that is, entities that generate values of `OptionalSubscriptionsFilter`, not the values themselves.  `NoFilter` happens to be a constructor for zero arity, which takes no arguments,  but that is just fine. Scala reuses OO keyword terminology for functional concepts, but the semantics here are that of ADT, so `case object` really means ADT constructor here.

When folding an option over ADT we must provide type annotation because of https://stackoverflow.com/a/29587749/5205022